### PR TITLE
Add support for editing autosde cloud volume

### PR DESCRIPTION
**Pull request by llivne**
This PR is part of PRs that solve Autosde cloud volume editing

image

If specific info about a cloud volume is passed then we are under update mode (not creation) and the update field should be used.

Links
----------------
https://github.com/ManageIQ/manageiq-ui-classic/pull/7788
https://github.com/ManageIQ/manageiq-providers-autosde/pull/72

### CODE REVIEW

**Comment by agrare** [comment]
Hey @llivne sorry for the delay on reviewing this I've been working on how we want to do the ddf params for editing records, so far we've only done `OPTIONS /api/collection`

I've added the ability to call `OPTIONS /api/collection/:id` which I think is a more natural way to do this.  So your API call will look like `API.options('/api/cloud_volumes/${recordId}')`

Then in this method you can handle this by checking for `params["id"]` to see if it was called on an instance rather than the collection.  You can then lookup the record by `resource_search()` and call an instance method `record.params_for_update` and not have to pass any params in.

File: app/controllers/api/cloud_volumes_controller.rb
Comment (agrare) [inline app/controllers/api/cloud_volumes_controller.rb:62]: I'm thinking this method can look like this now:
```
def options
  if params[:ems_id]
    ems = resource_search(params[:ems_id], :ext_management_systems, ExtManagementSystem)
    raise BadRequestError, "No CloudVolume support for - #{ems.class}" unless defined?(ems.class::CloudVolume)

    klass = ems.class::CloudVolume
    raise BadRequestError, "No DDF specified for - #{klass}" unless klass.respond_to?(:params_for_create)

    render_options(:cloud_volumes, :form_schema => klass.params_for_create(ems))
  elsif params[:id]
    cloud_volume = resource_search(params[:id], :cloud_volumes, CloudVolume)
    render_options(:cloud_volumes, :form_schema => cloud_volume.params_for_update)
  else
    super
  end
end
```

Comment (llivne) [reply app/controllers/api/cloud_volumes_controller.rb]: @agrare thank you for your comments. I updated all of the PRs to work like you suggested. please review the new changes

File: app/controllers/api/cloud_volumes_controller.rb
Comment (agrare) [inline app/controllers/api/cloud_volumes_controller.rb:60]: Prefer to check for `klass.supports?(:create)` rather than checking for methods existing.

```suggestion
        raise BadRequestError, klass.unsupported_reason(:create) unless klass.supports?(:create)
```

**Comment by agrare** [comment]
Looks good, can you add a spec test for these two new code paths?  You can use https://github.com/ManageIQ/manageiq-api/pull/1058/files#diff-a0d92b93db6899bc0858bc8ccfbcec74b815ab90846b58784bc0de9acdefdee2R101-R129 as a reference

File: app/controllers/api/cloud_volumes_controller.rb
Comment (llivne) [reply app/controllers/api/cloud_volumes_controller.rb]: done

**Comment by llivne** [comment]
> Looks good, can you add a spec test for these two new code paths? You can use https://github.com/ManageIQ/manageiq-api/pull/1058/files#diff-a0d92b93db6899bc0858bc8ccfbcec74b815ab90846b58784bc0de9acdefdee2R101-R129 as a reference

done :)

File: spec/requests/cloud_volumes_spec.rb
Comment (agrare) [inline spec/requests/cloud_volumes_spec.rb:11]: This looks like a typo

**Comment by agrare** [comment]
@bdunne can you also review?

File: spec/requests/cloud_volumes_spec.rb
Comment (bdunne) [inline spec/requests/cloud_volumes_spec.rb:221]: Please don't specify `:name`, it isn't important to your test and the factory will automatically create one for you

File: spec/requests/cloud_volumes_spec.rb
Comment (bdunne) [inline spec/requests/cloud_volumes_spec.rb:225]: Do you need this extra lookup?  It's the same as `volume` above, correct?

**Review by bdunne — DISMISSED** [review]

File: spec/requests/cloud_volumes_spec.rb
Comment (bdunne) [reply spec/requests/cloud_volumes_spec.rb:11]: Hmm... somehow Travis passed 😕

Comment (bdunne) [reply spec/requests/cloud_volumes_spec.rb:11]: Well, it's [a real thing](https://relishapp.com/rspec/rspec-core/v/3-10/docs/filtering/filter-run-when-matching#groups-can-be-focused-with-%60fdescribe%60-or-%60fcontext%60), but I don't think it was intentional.

File: spec/requests/cloud_volumes_spec.rb
Comment (llivne) [reply spec/requests/cloud_volumes_spec.rb]: fixed

File: spec/requests/cloud_volumes_spec.rb
Comment (llivne) [reply spec/requests/cloud_volumes_spec.rb]: you are correct. fixed

File: spec/requests/cloud_volumes_spec.rb
Comment (llivne) [reply spec/requests/cloud_volumes_spec.rb]: @bdunne @agrare sorry about that. This actually was intentional, but it wasnt supposed to be part of the PR. in order to run specific specs we sometime use[ Inclusion filters](https://relishapp.com/rspec/rspec-core/v/2-6/docs/filtering/inclusion-filters). this is because running a specific file not always works. in short, I wrote a script that add  config.filter_run focus: true to spec_helper.rb (simply sed -i '/RSpec.configure.*/a \ \ config.filter_run focus: true' ./spec/spec_helper.rb) and then only tagged tests will be run. Tagging is done by adding ':focus => true' (or prepend an f) to it, specify, describe or context.
my script also remove the modified spec_helper.rb when it done but not the tagging and I forgot to remove it before the push. this shouldn't append, sorry about that.

Comment (llivne) [reply spec/requests/cloud_volumes_spec.rb]: I removed it now :)

**Comment by llivne** [comment]
@bdunne @agrare I did all the required changes, waiting for your review/approval

**Comment by ereztheodorou** [comment]
@miq-bot add_label morphy/yes?

File: spec/requests/cloud_volumes_spec.rb
Comment (bdunne) [inline spec/requests/cloud_volumes_spec.rb:226]: Do we need to stub these?

Comment (llivne) [reply spec/requests/cloud_volumes_spec.rb]: @bdunne `allow(CloudVolume).to receive(:find).with(cloud_volume.id.to_s).and_return(cloud_volume)` is not needed so I removed it. the Rbac is required for the OPTIONS to work

**Review by agrare — APPROVED** [review]

**Comment by Fryguy** [comment]
Backported to `morphy` in commit d762bdf1ac80bdc6b5cf54d9d3509ea51e13024b.

```text
commit d762bdf1ac80bdc6b5cf54d9d3509ea51e13024b
Author: Adam Grare <adam@grare.com>
Date:   Tue Sep 7 09:12:18 2021 -0400

    Merge pull request #1056 from Autosde/fix_edit_cloud_volume
    
    Add support for editing autosde cloud volume
    
    (cherry picked from commit e6fa13e6e942051d8943e013538fdfb3c6a417d2)
```

**Pull request merged**
